src/main/java/com/shapesecurity/salvation/Parser.java

```diff
@@ -366,67 +548,154 @@ private boolean eat(@Nonnull Class<? extends Token> c) {
     }
 
-    @Nonnull private Set<SandboxValue> parseSandboxTokenList() throws ParseException {
+    @Nonnull private Set<ReferrerValue> parseReferrerTokenList() throws DirectiveParseException {
+
+        Set<ReferrerValue> referrerTokens = new LinkedHashSet<>();
+        if (this.hasNext(DirectiveValueToken.class)) {
+            boolean parseException = false;
+            Token dv = this.advance();
+            for (SubDirectiveValueToken subdv : splitByWSP(dv)) {
+                try {
+                    referrerTokens.add(this.parseReferrerToken(subdv));
+                } catch (DirectiveValueParseException e) {
+                    parseException = true;
+                }
+            }
+            if (parseException) {
+                throw INVALID_REFERRER_TOKEN_LIST;
+            }
+        }
+        if (referrerTokens.size() != 1) {
+            this.error("The referrer directive must contain exactly one referrer-token");
+            throw INVALID_REFERRER_TOKEN_LIST;
+        }
+        return referrerTokens;
+    }
+
+    @Nonnull private ReferrerValue parseReferrerToken(@Nonnull SubDirectiveValueToken token)
+        throws DirectiveValueParseException {
+        Matcher matcher = Constants.referrerTokenPattern.matcher(token.value);
+        if (matcher.find()) {
+            return new ReferrerValue(token.value);
+        }
+        this.error(token, "Expecting referrer-token but found " + token.value);
+        throw INVALID_REFERRER_TOKEN;
+    }
+
+    @Nonnull private Set<SandboxValue> parseSandboxTokenList() throws DirectiveParseException {
         Set<SandboxValue> sandboxTokens = new LinkedHashSet<>();
-        while (this.hasNext(DirectiveValueToken.class)) {
-            sandboxTokens.add(this.parseSandboxToken());
+        if (this.hasNext(DirectiveValueToken.class)) {
+            boolean parseException = false;
+            Token dv = this.advance();
+            for (SubDirectiveValueToken subdv : splitByWSP(dv)) {
+                try {
+                    sandboxTokens.add(this.parseSandboxToken(subdv));
+                } catch (DirectiveValueParseException e) {
+                    parseException = true;
+                }
+            }
+            if (parseException) {
+                throw INVALID_SANDBOX_TOKEN_LIST;
+            }
         }
         return sandboxTokens;
     }
 
-    @Nonnull private SandboxValue parseSandboxToken() throws ParseException {
-        Token token = this.advance();
-        Matcher matcher = Constants.sandboxTokenPattern.matcher(token.value);
+    @Nonnull private SandboxValue parseSandboxToken(@Nonnull SubDirectiveValueToken token) throws DirectiveValueParseException {
+        Matcher matcher = Constants.sandboxEnumeratedTokenPattern.matcher(token.value);
         if (matcher.find()) {
             return new SandboxValue(token.value);
+        } else {
+            this.warn(token, "The sandbox directive should contain only allow-forms, allow-modals, "
```

Comment:
How is this not an error? And if we know all of the possible values, shouldn't `SandboxValue` be an `enum` now?